### ISSUE TITLE
Azure Availability Set for VM Managed Disks

### ISSUE THREAD

**Issue by ghost** [rancher/rancher#15788]
**Rancher versions:**
rancher/rancher: **v2.0.8**

**Type/provider of hosts:** 
Azure

**Steps to Reproduce:**

Deploy Azure VM using Azure cloud provider

Size: **Standard_D2s_V3**
Storage Type: **Premium LRS**
Availability Set: **massimo-avs**

Fill out the following in the azure cloud provider config

aadClientId*
aadClientSecret*
subscriptionId*
tenantId*

**Results:**

Rancher deploys Azure VM which does not use managed disks, even though I specified Premium LRS as the storage type and my own availability set.
```
'massimo1' is not using Managed Disks. Migrate to Managed Disks to get more benefits.
```
The odd thing is that if you create the availability set in Azure Portal as managed before deploying, Rancher will change it to unmanaged on the deployment of the Azure VM.

Am I missing a configuration option that allows me to set the availability set to managed on the node template for Azure? If not then can we implement an option to select Unmanaged (classic) vs Managed (Aligned) availability set?

https://docs.microsoft.com/en-us/rest/api/compute/availabilitysets/createorupdate#request-body
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/tutorial-availability-sets#create-an-availability-set

**Comment by erikwilson** [comment]
Notes: Testing on master initially produced an error like:
``
2019/01/21 16:09:11 [ERROR] NodeController c-jn98d/m-2rfl2 [node-controller] failed with : Error creating machine: Error in driver during machine creation: compute.VirtualMachinesClient#CreateOrUpdate: Failure responding to request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=409 Code="OperationNotAllowed" Message="Addition of a VM with managed disks to non-managed Availability Set or addition of a VM with blob based disks to managed Availability Set is not supported. Please create an Availability Set with 'Aligned' SKU in order to add a VM with managed disks to it."
``

Followed by:
``
2019/01/21 16:12:44 [ERROR] NodeController c-jn98d/m-5298x [node-controller] failed with : Error creating machine: Error in driver during machine creation: compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=200 -- Original Error: Long running operation terminated with status 'Failed': Code="VMSizeDoesntSupportPremiumStorage" Message="Requested operation cannot be performed because storage account type 'Premium_LRS' is not supported for VM size 'Standard_A2'."
``

Changed size to `Standard_DS11_v2` and now getting errors like:
``
2019/01/21 16:57:10 [ERROR] ClusterController c-kcs7m [cluster-provisioner-controller] failed with : Can not find RKE state file: open /Users/erik/go/src/github.com/rancher/rancher/management-state/rke/rke-867082529/cluster.rkestate: no such file or directory
``

**Comment by erikwilson** [comment]
After creating a new 'managed' availability set in the Azure portal and then creating a new cluster using that availability set I am seeing errors like:
``
2019/01/22 10:07:47 [ERROR] NodeController c-cfmv5/m-xfllm [node-controller] failed with : Error creating machine: Error in driver during machine creation: compute.VirtualMachinesClient#CreateOrUpdate: Failure responding to request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=409 Code="OperationNotAllowed" Message="Addition of a VM with managed disks to non-managed Availability Set or addition of a VM with blob based disks to managed Availability Set is not supported. Please create an Availability Set with 'Aligned' SKU in order to add a VM with managed disks to it."
``

However, after some period of time Rancher attempts to re-provision the nodes and is successful provisioning using that availability set. Further clusters are able to be provisioned without encountering the same error. This indicates there may be an issue with the Azure API or one of the supporting libraries.

On master (1444ef028a)

**Comment by cjellick** [comment]
@erikwilson can you test two things:
1. if it is a timing issue where the managed AS needs to exist for some time before it works in their API
2. if this works in the latest 2.1 release

**Comment by erikwilson** [comment]
Sorry, actually it isn't working properly on master either. It looks like we might be deleting and re-creating the availability set as not managed. I am looking into why the initial `docker-machine create` failed, but we may want to add a managed flag for the UI when creating availability sets.

**Comment by erikwilson** [comment]
It looks like in order to support managed disks in availability sets properly that `docker-machine` will need to be modified to include SDK ManagedDisk parameters first introduced in 8.1.0-beta of github.com/Azure/azure-sdk-for-go (currently 5.0.0-beta). Unfortunately backwards compatibility issues prevent upgrading azure-sdk-for-go directly to 8.1.0-beta.

**Comment by danielllek** [comment]
When this could be resolved?

I got one step further into the pit: 
* I've upgraded availability set and converted to managed disk (needed more iops), 
* now I can't add another node via rancher (rancher tries to to add vm with unmanaged disk to "managed" availability set which is not possible)
* "managed" availability set can't be downgraded

Now I'm stuck...

**Comment by sirrapa** [comment]
We encounter the same issues @danielllek mentioned.
This is a blocker for our company to use Rancher because we have to use managed disks.
A solution or work-arround is very, very, very welcome...

**Comment by joosthof** [comment]
@deniseschannon @erikwilson This is a major blocker for our company, since we can only go live with managed disks. Can this bug get priority on the backlog? Is there a work-around available?

**Comment by fabionitto** [comment]
@deniseschannon @erikwilson I'm also having this problem at my company. We had to convert the availability set to ManagedDisk, so that all workers could have its disks migrated to. This was done because Managed Disks created by StorageClasses couldn't be attached to workers.

Now we can't scale workers because of this availability set limitation.

**Comment by fabionitto** [comment]
> @erikwilson can you test two things:
> 
> 1. if it is a timing issue where the managed AS needs to exist for some time before it works in their API
> 2. if this works in the latest 2.1 release

I'm using Rancher v2.2.2 and this is not a timing issue here...

**Comment by mlushpenko** [comment]
Same problem here and I can't find any option in rancher to change this

**Comment by cjellick** [comment]
backend is in.

**Comment by annerajb** [comment]
Just stumbled upon this issue. 
Is there a rancher 2 release that contains this even at an alpha state?
Or a workaround to use the new backend?

**Comment by rmweir** [comment]
@annerajb this depends on an update to docker-machine that is not included in a current rc. Once it is included, you should be able to do this by using the UI as the fields are generated dynamically. This should be available in 2.3. I can ask about changing the milestone.

**Comment by jiaqiluo** [comment]
The bug is reproduced in Rancher: v2.2.7

Deploy Azure VM using Azure cloud provider

Test 1: let Rancher generates all resources 

Size: **Standard_D2s_V3**
Storage Type: **Premium LRS**
Availability Set: jiaqi-availablity-set1 (**created by Rancher**)

Result:
- Rancher deploys Azure VM which does not use managed disks
<img width="1366" alt="Screen Shot 2019-08-08 at 4 10 39 PM" src="https://user-images.githubusercontent.com/6218999/62744266-abfda480-b9fa-11e9-8245-f2026fd5fdde.png">

Test 2: use my own availability set 
- create my own availability set and set it to use managed disk
<img width="1370" alt="Screen Shot 2019-08-08 at 4 12 21 PM" src="https://user-images.githubusercontent.com/6218999/62744319-ecf5b900-b9fa-11e9-964c-2584f51167ef.png">
- use this availability set to provision a cluster

Size: **Standard_D2s_V3**
Storage Type: **Premium LRS**
Availability Set: jiaqi-managed-disk-a-set


Result:

- Still, Rancher deploys Azure VM which does not use managed disks
- Rancher modifies my availability set to not use managed disk
<img width="1368" alt="Screen Shot 2019-08-08 at 4 30 27 PM" src="https://user-images.githubusercontent.com/6218999/62744344-14e51c80-b9fb-11e9-9538-00edc18953f4.png">

**Comment by jiaqiluo** [comment]
The bug fix is validated in Rancher: master-head bc3700f8

Test 1: let Rancher generates all resources 

Size: **Standard_D2s_V3**
Storage Type: **Premium LRS**
Availability Set: jiaqi-from-rancher (**created by Rancher**)

Result:	
- Rancher deploys Azure VM which are using managed disks
<img width="1380" alt="Screen Shot 2019-08-09 at 10 00 35 AM" src="https://user-images.githubusercontent.com/6218999/62796004-e7e14a00-ba8c-11e9-9d80-ae439986439a.png">

- Rancher creates the availability set that set the `managed` to yes which means it uses managed disks
<img width="1375" alt="Screen Shot 2019-08-09 at 10 01 16 AM" src="https://user-images.githubusercontent.com/6218999/62796093-27a83180-ba8d-11e9-8967-731a2512a8b8.png">

Test 2: use my own availability set 
- create my own availability set and set it to use managed disk

<img width="1376" alt="Screen Shot 2019-08-08 at 5 22 44 PM" src="https://user-images.githubusercontent.com/6218999/62796138-4f979500-ba8d-11e9-910b-4c3dd07abac9.png">

- use this availability set to provision a cluster

Size: **Standard_D2s_V3**
Storage Type: **Premium LRS**
Availability Set: jiaqi-managed-disk-a-set


Result:
- Rancher deploys Azure VM which are using managed disks
<img width="1377" alt="Screen Shot 2019-08-08 at 5 24 04 PM" src="https://user-images.githubusercontent.com/6218999/62796188-6fc75400-ba8d-11e9-8aca-9b88b3d3a94f.png">
